File: csharp/1524.number-of-sub-arrays-with-odd-sum.cs
/*
 * @lc app=leetcode id=1524 lang=csharp
 *
 * [1524] Number of Sub-arrays With Odd Sum
 */

// @lc code=start
public class Solution {
    public int NumOfSubarrays(int[] arr) {
        int ret = 0, even = 0, odd = 0;
        foreach (int x in arr) {
            if (x % 2 == 0) {
                ++even;
            } else {
                int tmp = odd;
                odd = even + 1;
                even = tmp;
            }
            ret = (ret + odd) % 1000000007;
        }
        return ret;
    }
}
// @lc code=end

File: java/1524.number-of-sub-arrays-with-odd-sum.java
/*
 * @lc app=leetcode id=1524 lang=java
 *
 * [1524] Number of Sub-arrays With Odd Sum
 */

// @lc code=start
class Solution {
    public int numOfSubarrays(int[] arr) {
        int acc = 0, ret = 0, odds = 0, n = arr.length;
        for (int i = 0; i < n; ++i) {
            acc ^= (arr[i] & 1);
            ret = (ret + odds * (1 - acc) + (1 + i - odds) * acc) % 1000000007;
            odds += acc;
        }
        return ret;
    }
}
// @lc code=end

